compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -98,52 +118,91 @@ createCond :: BoolExpr
+-- TODO: This can be extended with additional cases in the future.
+simplifyCond :: Eq a => Cond a -> [Cond a]
+simplifyCond (Conditional (BAnd b1 b2) xs ys)
+  | b1 == b2 = simplifyCond (Conditional b1 xs ys)
+  | b1 == BNot b2 = ys
+  | b2 == BNot b1 = ys
+simplifyCond (Conditional b [Conditional b1 xs1 _] [Conditional b2 _ ys2])
+  | b1 == b2 = simplifyCond (Conditional b xs1 ys2)
+simplifyCond (Conditional _ xs ys)
+  | xs == ys = concatMap simplifyCond xs
+simplifyCond c = [c]
+
+-- | Shift the conditional inside of the update set, by extending each update
+-- with the condition.
+introCond :: IsPhase ph => Cond (UpdateSet ph) -> UpdateSet ph
 introCond (Determined upds) = upds
-introCond (Conditional e updx updy) = case getPhase updx of
-  UpdateSetVG{} -> UpdateSetVG
-    (buildCond updx updy _usvgUpdate)
-    (buildCond updx updy _usvgChoice)
-    (buildCond updx updy _usvgValue)
-  UpdateSetCG{} -> UpdateSetCG
-    (buildCond updx updy _uscgUpdate)
-    (buildCond updx updy _uscgChoice)
-  UpdateSetS{} -> UpdateSetS
-    (buildCond updx updy _ussUpdate)
+introCond (Conditional e updx updy) = buildCond e updx updy extCondUpd
   where
-    -- | Construct a single conditional, if the input is not empty.
-    buildCond :: [Cond (UpdateSet ph)]
-      -- ^ The input for the true case.
+    -- | Construct a single conditional update set, combining the two input lists
+    -- and the boolean expression, if the input is non-empty.
+    buildCond :: IsPhase ph
+      => BoolExpr
       -> [Cond (UpdateSet ph)]
-      -- ^ The input for the false case.
-      -> (UpdateSet ph -> [Cond a])
-      -- ^ The fetch function.
-      -> [Cond a]
-    buildCond updx updy get =
-      let xs = concatCond updx get
-          ys = concatCond updy get
-      in [Conditional e xs ys | not (null xs && null ys)]
-
-    -- TODO: Temporary solution. Make introCond a part of the GenPhase class instead.
-    getPhase :: [Cond (UpdateSet ph)] -> UpdateSet ph
-    getPhase lst = case head lst of
-      Determined upds -> upds
-      Conditional _ xs _ -> getPhase xs
-
--- | Fetch the conditionals from the conditional update set, and flatten the
--- two layers into one.
-concatCond :: [Cond (UpdateSet ph)]
-  -- ^ The conditional update set to fetch from.
-  -> (UpdateSet ph -> [Cond a])
-  -- ^ The fetch function.
-  -> [Cond a]
-concatCond upds get =
-  let upds' = map introCond upds
-  in concatMap get upds'
-
--- | Data type denoting an update.
-data Upd
+      -> [Cond (UpdateSet ph)]
+      -> (BoolExpr -> Upd ph -> UpdateSet ph)
+      -> UpdateSet ph
+    buildCond bexp cxs cys ext =
+      let xs = concatMap introCond cxs
+          ys = concatMap introCond cys
+      in concatMap (ext bexp) xs ++ concatMap (ext $ BNot bexp) ys
+
+-- | Data type denoting a potential recursion cycle.
+data Rec a
+  = Simple a
+  -- ^ Basic, non-recursive value.
+  | Rec [a]
+  -- ^ (Possibly multiple) recursion cycles.
+  | MutRec [(String,a)]
```

Comment:
At the moment, the names of every let value / choice included within the mutual recursion cycle. But this could easily be extended with additional debugging information in the future. Right now, a simple String seemed sufficient.